pkg/controller/promotionreconciler/prowjobretrigger/reconciler.go

```diff
@@ -0,0 +1,173 @@
+package prowjobretrigger
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/workqueue"
+	prowjobsv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/github"
+	"sigs.k8s.io/controller-runtime"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	cioperatorapi "github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/controller/promotionreconciler/prowjobreconciler"
+	controllerutil "github.com/openshift/ci-tools/pkg/controller/util"
+	"github.com/openshift/ci-tools/pkg/load/agents"
+	"github.com/openshift/ci-tools/pkg/rehearse"
+)
+
+type Options struct {
+	CIOperatorConfigAgent agents.ConfigAgent
+	GitHubClient          github.Client
+}
+
+const (
+	ControllerName = "promotion_job_retrigger"
+)
+
+func AddToManager(mgr controllerruntime.Manager, opts Options, enqueuer prowjobreconciler.Enqueuer) error {
+	log := logrus.WithField("controller", ControllerName)
+	r := &reconciler{
+		ctx:          context.Background(),
+		log:          log,
+		client:       mgr.GetClient(),
+		gitHubClient: opts.GitHubClient,
+		enqueueJob:   enqueuer,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler:              r,
+		MaxConcurrentReconciles: 10,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+	if err := c.Watch(
+		&source.Kind{Type: &prowjobsv1.ProwJob{}},
+		handler.Funcs{
+			UpdateFunc: func(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+				q.Add(e)
```

Comment:
While this compiles ts not going to work at all. You must put `reconcile.Request` in the q and not events, otherwise your controller will never do anything except log errors. I recommend using `hack/run-dptp-controller-manager.sh`  btw to test the basic functionality

Reply:
I like to rely on types to be my crutch for not making dumb mistakes :)